tests/gnrc_sock_tcp/main.c

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright (C) 2021   Simon Brummer <simon.brummer@posteo.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+#include <stdio.h>
+#include <string.h>
+
+#include "shell.h"
+#include "msg.h"
+#include "net/sock/tcp.h"
+
+#define MAIN_QUEUE_SIZE (8)
+#define SOCK_TCP_QUEUE_SIZE (1)
+#define BUFFER_SIZE (1024)
+
+static msg_t main_msg_queue[MAIN_QUEUE_SIZE];
+static sock_tcp_t socks[SOCK_TCP_QUEUE_SIZE];
+static sock_tcp_t *sock = socks;
+static sock_tcp_queue_t sock_queue;
+static char buffer[BUFFER_SIZE];
+
+void dump_args(int argc, char **argv)
+{
+    printf("%s: ", argv[0]);
+    printf("argc=%d", argc);
+    for (int i = 0; i < argc; ++i) {
+        printf(", argv[%d] = %s", i, argv[i]);
+    }
+    printf("\n");
+}
+
+void print_result(const char *name, int err)
+{
+    if (err) {
+        printf("%s: returns %s\n", name, strerror(err));
+    } else {
+        printf("%s: returns %d\n", name, err);
+    }
+}
+
+int sock_tcp_connect_cmd(int argc, char **argv)
+{
+    dump_args(argc, argv);
+
+    sock_tcp_ep_t ep = SOCK_IPV6_EP_ANY;
+    gnrc_tcp_ep_from_str((gnrc_tcp_ep_t *) &ep, argv[1]);
+    uint16_t local_port = atol(argv[2]);
+    uint16_t flags = 0;
+
+    int err = sock_tcp_connect(sock, &ep, local_port, flags);
+    print_result(argv[0], err);
+    return 0;
+}
+
+int sock_tcp_disconnect_cmd(int argc, char **argv)
+{
+    dump_args(argc, argv);
+    sock_tcp_disconnect(sock);
+    printf("%s: returns\n", argv[0]);
+    return 0;
+}
+
+int sock_tcp_listen_cmd(int argc, char **argv)
+{
+    dump_args(argc, argv);
+    sock_tcp_ep_t ep = SOCK_IPV6_EP_ANY;
+    gnrc_tcp_ep_from_str((gnrc_tcp_ep_t *) &ep, argv[1]);
+    uint16_t flags = 0;
+
+    int err = sock_tcp_listen(&sock_queue, &ep, socks, SOCK_TCP_QUEUE_SIZE, flags);
+    print_result(argv[0], err);
+    return 0;
+}
+
+int sock_tcp_stop_listen_cmd(int argc, char **argv)
+{
+    dump_args(argc, argv);
+    sock_tcp_stop_listen(&sock_queue);
+    printf("%s: returns\n", argv[0]);
+    return 0;
+}
+
+int sock_tcp_accept_cmd(int argc, char **argv)
+{
+    dump_args(argc, argv);
+    sock_tcp_t *tmp = NULL;
+    uint16_t timeout = atol(argv[1]);
+
+    int err = sock_tcp_accept(&sock_queue, &tmp, timeout);
+    print_result(argv[0], err);
+
+    if (tmp) {
+        sock = tmp;
+    }
+    return 0;
+}
+
+int sock_tcp_read_cmd(int argc, char **argv)
+{
+    dump_args(argc, argv);
+
+    int timeout = atol(argv[1]);
+    unsigned to_receive = atol(argv[2]);
```

Comment:
Would be nice if those were swapped

```suggestion
    int timeout = argv > 1 ? atol(argv[2]) : 0;
    unsigned to_receive = atol(argv[1]);
```

But I actually get stuck in a loop here if I try to read more than what's in the buffer:

```
> sock_tcp_read 100 10
sock_tcp_read: argc=3, argv[0] = sock_tcp_read, argv[1] = 100, argv[2] = 10
sock_tcp_read: received 10 It's worki
> sock_tcp_read 0 10
sock_tcp_read: argc=3, argv[0] = sock_tcp_read, argv[1] = 0, argv[2] = 10
sock_tcp_read: returns Unknown error -11
sock_tcp_read: returns Unknown error -11
sock_tcp_read: returns Unknown error -11
```

Reply:
Without a timeout, you basically trigger non-blocking behavior. If no data is available, -EAGAIN ist returned. This is not a fault of the implementation, the test just tries to continue to read until the required amount of data is available. The test assumes that amount of data that should be read, is actually sent.

Reply:
If you want to return here, you have to specify a non-zero timeout.

Reply:
Ok, that makes sense.

Reply:
Doesn't this contradict the description of `SOCK_NO_TIMEOUT` though?

Reply:
Hmm you might have a point here. I'll need to think about it in detail after my working hours ;)